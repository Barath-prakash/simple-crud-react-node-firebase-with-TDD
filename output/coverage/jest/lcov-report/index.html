<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for All files</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="prettify.css" />
    <link rel="stylesheet" href="base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      All files
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">72.68% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>133/183</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">57.89% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>55/95</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">71.76% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>61/85</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">73.21% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>123/168</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line medium'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="src"><a href="src/index.html">src</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="src/action"><a href="src/action/index.html">src/action</a></td>
	<td data-value="88.57" class="pic high"><div class="chart"><div class="cover-fill" style="width: 88%;"></div><div class="cover-empty" style="width:12%;"></div></div></td>
	<td data-value="88.57" class="pct high">88.57%</td>
	<td data-value="35" class="abs high">31/35</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="20" class="abs high">16/20</td>
	<td data-value="84" class="pct high">84%</td>
	<td data-value="25" class="abs high">21/25</td>
	</tr>

<tr>
	<td class="file high" data-value="src/components/common"><a href="src/components/common/index.html">src/components/common</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

<tr>
	<td class="file medium" data-value="src/components/forms/employees"><a href="src/components/forms/employees/index.html">src/components/forms/employees</a></td>
	<td data-value="60.61" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 60%;"></div><div class="cover-empty" style="width:40%;"></div></div></td>
	<td data-value="60.61" class="pct medium">60.61%</td>
	<td data-value="66" class="abs medium">40/66</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="60" class="abs medium">30/60</td>
	<td data-value="45.45" class="pct low">45.45%</td>
	<td data-value="22" class="abs low">10/22</td>
	<td data-value="64.52" class="pct medium">64.52%</td>
	<td data-value="62" class="abs medium">40/62</td>
	</tr>

<tr>
	<td class="file medium" data-value="src/components/pages/employees"><a href="src/components/pages/employees/index.html">src/components/pages/employees</a></td>
	<td data-value="74.29" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 74%;"></div><div class="cover-empty" style="width:26%;"></div></div></td>
	<td data-value="74.29" class="pct medium">74.29%</td>
	<td data-value="70" class="abs medium">52/70</td>
	<td data-value="70.37" class="pct medium">70.37%</td>
	<td data-value="27" class="abs medium">19/27</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="40" class="abs high">32/40</td>
	<td data-value="75.36" class="pct medium">75.36%</td>
	<td data-value="69" class="abs medium">52/69</td>
	</tr>

<tr>
	<td class="file medium" data-value="src/reducers"><a href="src/reducers/index.html">src/reducers</a></td>
	<td data-value="71.43" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 71%;"></div><div class="cover-empty" style="width:29%;"></div></div></td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="8" class="abs medium">6/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	</tr>

<tr>
	<td class="file high" data-value="src/shared/commonMessages"><a href="src/shared/commonMessages/index.html">src/shared/commonMessages</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri May 03 2019 16:36:50 GMT+0530 (India Standard Time)
</div>
</div>
<script src="prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="sorter.js"></script>
<script src="block-navigation.js"></script>
</body>
</html>
