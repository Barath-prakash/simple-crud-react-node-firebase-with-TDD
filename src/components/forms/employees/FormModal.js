import React, { useState, useEffect } from 'react';
import { Button } from 'react-bootstrap';
import { Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';
import moment from 'moment';
import Datepicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import Validator from 'validator';
import { MSG } from 'shared/commonMessages/messages';
moment.suppressDeprecationWarnings = true;

const FormModal = ({ employeeFormModal, employee, addEmployee, updateEmployee, employeeFormMoalClose }) => {

    let stateObj = {
        data: {
            empId: "",
            firstName: "",
            lastName: "",
            email: "",
            gender: "",
            dateOfBirth: "",
            age: ""
        },
        errors: {},
        loading: false
    };
   
    const [state, setState] = useState(stateObj);
    
    useEffect(() => {
        if(Object.keys(employee).length > 0) {
            let stateObjVal = {
                data: {
                    empId: employee.empId,
                    firstName: employee.firstName,
                    lastName: employee.lastName,
                    email: employee.email,
                    gender: employee.gender,
                    dateOfBirth: employee.dateOfBirth,
                    age: employee.age
                },
                errors: {},
                loading: false
            };
            setState(stateObjVal);
        }
    }, [employee]);

        const handleCloseFormModal = () => {
            resetForm();
            employeeFormMoalClose();
        }
    
        const handleChange = (e) => {
            // this.setState({ data: {...this.state.data, [e.target.name]: e.target.value } });
            setState({...state, data: { ...data, [e.target.name] : e.target.value } });
        }
        
        const handleDateChange = (date) => {
            if(date !== null && date !== "Invalid date") {
                setState({...state, data: { ...data, dateOfBirth : date, age: moment().diff(moment(date).format('L'), 'years') } });
            } else {
                setState({...state, data: { ...data, dateOfBirth : "", age: "" } });
            }
        }
        
        const handleSubmit = (e) => {
            e.preventDefault();
            let data = state.data;
            const errors = validate(data);
            setState({ ...state, errors });
            if(Object.keys(errors).length === 0) {
                setState({ ...state, loading: true })
                if(state.data.empId) {
                    return updateEmployee(data).then(res => handleCloseFormModal()).catch(err => setState({ ...state, loading: false }))
                } else {
                    return addEmployee(data).then(res => handleCloseFormModal()).catch(err => setState({ ...state, loading: false }))
                }
            }
        }
    
        const validate = (data) => {
            const errors = {};
    
            if(Validator.isEmpty(data.firstName)) {
               errors['firstName'] = MSG['firstName_cannot_empty'];
            } else if(!Validator.isLength(data.firstName, { min: 3 })) {
               errors['firstName'] = MSG['firstName_length'];
            }
    
            if(Validator.isEmpty(data.lastName)) {
                errors['lastName'] = MSG['lastName_cannot_empty'];
            } else if(!Validator.isLength(data.lastName, { min: 2 })) {
                errors['lastName'] = MSG['lastName_length'];
            }
    
            if(Validator.isEmpty(data.email)) {
                errors['email'] = MSG['email_cannot_empty'];
            } else if(!Validator.isEmail(data.email)) {
                errors['email'] = MSG['email_invalid'];
            }
    
            if(Validator.isEmpty(data.gender)) {
                errors['gender'] = MSG['gender_cannot_empty'];
            }
    
            if(Validator.isEmpty(data.dateOfBirth.toString().trim())) {
                errors['dateOfBirth'] = MSG['dob_cannot_empty'];
            }
    
            if(Validator.isEmpty(data.age.toString().trim()) || data.age === 0) {
                errors['age'] = MSG['age_cannot_empty'];
            }
    
            return errors;
        }
        
        const resetForm = () => {
            setState(stateObj);
        }

        const { data, loading, errors } = state;
        return (
            <Modal 
                id="modal"
                size="lg" 
                backdrop="static"
                isOpen={employeeFormModal} 
                toggle={() => handleCloseFormModal()} 
                >
                <ModalHeader toggle={() => handleCloseFormModal()}>
                    {data.empId ? "Update employee" : "Add employee"}
                </ModalHeader>
                <form onSubmit={(e) => handleSubmit(e)}>
                <ModalBody className="modal-overlay">
                    <div className="form-group">
                        <label htmlFor="firstName">First name:</label>
                        <input type="text" className="form-control" id="fn" name="firstName" value={data.firstName ? data.firstName: ""} onChange={(e) => handleChange(e)} placeholder="Enter first name" maxLength="75"/>
                        {errors.firstName && <span className="fld-error">{errors.firstName}</span>}
                    </div>
                    <div className="form-group">
                        <label htmlFor="lastName">Last name:</label>
                        <input type="text" className="form-control" id="ln" name="lastName" value={data.lastName ? data.lastName: ""} onChange={(e) => handleChange(e)} placeholder="Enter last name" maxLength="50"/>
                        {errors.lastName && <span className="fld-error">{errors.lastName}</span>}
                    </div>
                    <div className="form-group">
                        <label htmlFor="email">Email:</label>
                        <input type="email" className="form-control" id="em" name="email" value={data.email ? data.email: ""} onChange={(e) => handleChange(e)} placeholder="Enter email address" maxLength="200"/>
                        {errors.email && <span className="fld-error">{errors.email}</span>}
                    </div>
                    <div className="form-group">
                        <label htmlFor="email" id="gender">Gender:</label>
                            <label className="btn btn-default" style={{marginTop: 5}}>
                                <input type="radio" id="gender-male" name="gender" value="male" onChange={(e) => handleChange(e)} checked={data.gender && data.gender === "male"}/> Male
                            </label> 
                            <label className="btn btn-default" style={{marginTop: 5}}>
                                <input type="radio" id="gender-female" name="gender" value="female" onChange={(e) => handleChange(e)} checked={data.gender && data.gender === "female"}/> Female
                            </label><br/> 
                        {errors.gender && <span className="fld-error">{errors.gender}</span>}
                    </div>  
                    <div>
                    <label htmlFor="dateOfBirth">Date of birth:</label><br/>
                    <Datepicker
                        id="datepicker"
                        selected={(data.dateOfBirth && data.dateOfBirth !== "Invalid date") ? new Date(data.dateOfBirth) : null}
                        onChange={(e) => handleDateChange(e)}
                        peekNextMonth={true}
                        showMonthDropdown={true}
                        showYearDropdown={true}
                        dropdownMode="select"
                        placeholderText="DD-MM-YYYY"
                        dateFormat="dd-MM-yyyy"
                        shouldCloseOnSelect={true}
                        isClearable={true}
                        maxDate={new Date()}
                    />
                    <br/>
                    {errors.dateOfBirth && <span className="fld-error">{errors.dateOfBirth}</span>}
                    </div>
                    <div className="loading"></div>
                    <div className="form-group" style={{marginTop: 10}}>
                        <label htmlFor="age">Age:</label>
                        <input type="text" className="form-control" id="ag" value={data.age ? data.age: ""} maxLength="2" readOnly placeholder="Select your date of birth to fill this field"/>
                        {errors.age && <span className="fld-error">{errors.age}</span>}
                    </div>
                </ModalBody>
                <ModalFooter>
                    <Button id="cancelButton" type="button" variant="secondary" onClick={() => handleCloseFormModal()} disabled={loading}>Close</Button>
                    <Button id="submitButton" type="submit" variant="primary" disabled={loading}>Save</Button>
                </ModalFooter>
                </form> 
            </Modal>
        )
   }

export default FormModal;