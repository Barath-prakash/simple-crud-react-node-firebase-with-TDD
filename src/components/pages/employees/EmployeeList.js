import React, { useState } from 'react';
import { Table } from 'react-bootstrap';
import { Button } from 'react-bootstrap';
import moment from 'moment';

const EmployeeList = ({ employees, editEmployeeL, deleteEmployee, mDelete, mDeleteEmpId, deleteMultipleEmployees }) => {

    const [value, setValue] = useState({});

    const handleEmpChecked = (e) => {
        value[e.target.id] = e.target.checked;
        setValue(value);
        mDeleteEmpId({checked: e.target.checked, empId: e.target.id });
    }

    const tableBodyDiv = (
        (employees !== undefined && employees.length > 0) &&
        employees.map((employee, index) => {
        return (
            <tr key={index} id="dataRow">
                <td style={{textAlign: "center"}}>
                    <input className="chkInput" type="checkbox" checked={value[employee.empId] ? true : false} id={employee.empId} onChange={(e) => handleEmpChecked(e)}/>
                </td>
                <td>{employee.firstName +" "+ employee.lastName}</td>
                <td>{employee.email}</td>
                <td>{employee.gender}</td>
                <td>{moment(employee.dateOfBirth).format("DD-MM-YYYY")}</td>
                <td>{employee.age}</td>
                <td>
                <Button id="editBtn" variant="success" onClick={() => editEmployeeL(employee.empId)}>Edit</Button>
                <Button id="deleteBtn" variant="danger" style={{marginLeft: 5}} onClick={() => deleteEmployee(employee.empId)}>Delete</Button>
                </td>
            </tr>
        )
        })
    )

    const emptyDiv = <tr><td colSpan="7">No employees found</td></tr>
    
    return (
            <Table id="empList" striped bordered hover>
                <thead>
                <tr>
                    <th>Multiple delete</th>
                    <th>Employee name</th>
                    <th>Email</th>
                    <th>Gender</th>
                    <th>DOB</th>
                    <th>Age</th>
                    <th>Action</th>
                </tr>
                </thead>
                <tbody>
                    {(employees && employees.length > 0) ? tableBodyDiv : emptyDiv}
                    {mDelete &&
                    <tr>
                    <td style={{textAlign: "center"}}>
                        <Button  id="editBtn1" variant="danger" style={{marginLeft: 5}} onClick={() => deleteMultipleEmployees()}>Delete</Button>
                    </td>
                    <td colSpan="6"></td>
                    </tr>
                    }
                </tbody>
            </Table>
    )
}

export default EmployeeList;