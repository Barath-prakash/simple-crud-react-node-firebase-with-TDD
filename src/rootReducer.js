import { combineReducers } from 'redux';
import employees from 'reducers/employeesReducer';

const rootReducer = combineReducers({
    employees
});

export default rootReducer;