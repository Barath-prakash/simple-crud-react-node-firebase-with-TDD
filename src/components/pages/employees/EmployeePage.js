import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from 'react-bootstrap';
import { MSG } from 'shared/commonMessages/messages';

//React-Notification
import NotificationSystem from 'react-notification-system';

// Import dependent components
import EmployeeList from './EmployeeList';
import FormModal from 'components/forms/employees/FormModal';

// Action
import { fetchAllEmployeesA, employeeFormMoalOpenA, employeeFormMoalCloseA, 
         addEmployeeA, fetchEmployeeA, updateEmployeeA, deleteEmployeeA, deleteMultipleEmployeesA } from 'action/employeeAction';

const notifyStyle = {
    NotificationItem: { // Override the notification item
        DefaultStyle: { // Applied to every notification, regardless of the notification level
            margin: '100px 5px 5px 5px',
            padding: '20px 5px 15px 5px',
            fontSize: 17,
            textAlign: 'center'      
        }
    }
}
           
class EmployeePage extends Component {
    state = {
        mDeleteEmpArr: []
    }

    componentDidMount() {
        this.fetchAllEmployees();
    }

    // P
    fetchAllEmployees = () => {
        return this.props.fetchAllEmployees1();
    }
 
    // P
    employeeFormMoalOpen = () => {
        return this.props.employeeFormMoalOpen1();
    }

    employeeFormMoalClose = () => {
        return this.props.employeeFormMoalClose1();
    }

    addEmployee = (data) => {
        return this.props.addEmployee1(data).then(res => {
            this.fetchAllEmployees();
            this.setState({ mDeleteEmpArr: [] });
            this.notifyMsgFn(MSG['added_success'], 'success');
        })
    }

    editEmployeeP = (empId) => {
       return this.props.fetchEmployee1(empId).then(res => {
        this.employeeFormMoalOpen();
      })
    }

    updateEmployee = (data) => {
        return this.props.updateEmployee1(data).then(res => {
            this.fetchAllEmployees();
            this.setState({ mDeleteEmpArr: [] });
            this.notifyMsgFn(MSG['updated_success'], 'success');
        })
    }

    deleteEmployee = (empId) => {
        return this.props.deleteEmployee1(empId).then(res => {
            this.fetchAllEmployees();
            this.setState({ mDeleteEmpArr: [] });
            this.notifyMsgFn(MSG['removed_success'], 'success');
        })  
    }

    deleteEmployeeConfirm = (empId) => {
        let id = empId;
        this.notifyRef.addNotification({
            position: "tc",
            level: 'error',
            message: 'Are you sure you want to delete this employee?',
            autoDismiss: 0,
            action: {
                label: 'Ok',
                callback: () => {
                  this.deleteEmployee(id);
                }
              }
        })
    }

    mDeleteEmpId = (empData) => {
      let empIdArr = [...this.state.mDeleteEmpArr];
      if(empData.checked) {
        empIdArr.push(empData.empId);
      } else {
        let existIdInx = empIdArr.findIndex(data => data === empData.empId)
        if(existIdInx > -1) {
           empIdArr.splice(existIdInx, 1);
        }
      }
      this.setState({ mDeleteEmpArr: empIdArr });
    }

    deleteMultipleEmployeesConfirm = () => {
        this.notifyRef.addNotification({
            position: "tc",
            level: 'error',
            message: 'Are you sure you want to delete multiple employees?',
            autoDismiss: 0,
            action: {
                label: 'Ok',
                callback: () => {
                  this.deleteMultipleEmployeesC();
                }
              }
        })
        
    }

    deleteMultipleEmployeesC = () => {
        return this.props.deleteMultipleEmployees1(this.state.mDeleteEmpArr)
        .then(res => {
            this.setState({ mDeleteEmpArr: [] });
            this.fetchAllEmployees();
            this.notifyMsgFn(MSG['removed_success'], 'success');
        })
    }

    notifyMsgFn = (message, level) => {
        this.notifyMsg.addNotification({
            position: "tc",
            message: message,
            level: level
        });
    }

    render() {
        const { employees, employeeFormModal, employee } = this.props;
        const { mDeleteEmpArr } = this.state;
        return (
            <div className="container">
            <NotificationSystem ref={input => this.notifyRef = input} style={notifyStyle}/>
            <div>
                <h4 style={{float: 'left'}}>Employees List</h4>
                <Button variant="primary" style={{float: 'right'}} onClick={this.employeeFormMoalOpen}>Add employee</Button>
            </div><br/><br/>
                <EmployeeList 
                    id="employeeList"
                    employees={employees} 
                    editEmployeeL={this.editEmployeeP} 
                    deleteEmployee={this.deleteEmployeeConfirm} 
                    mDeleteEmpId={this.mDeleteEmpId} 
                    mDelete={mDeleteEmpArr.length > 1 ? true : false}
                    deleteMultipleEmployees={this.deleteMultipleEmployeesConfirm}
                    />
                <FormModal 
                    employeeFormMoalClose={this.employeeFormMoalClose} 
                    employeeFormModal={employeeFormModal}
                    employee={employee}
                    addEmployee={this.addEmployee} 
                    updateEmployee={this.updateEmployee}
                    />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    let employees = state.employees.employees ? state.employees.employees : [];
    let employeeFormModal = state.employees.employeeFormModal ? state.employees.employeeFormModal : false;
    let employee = state.employees.employee ? state.employees.employee : {};
    return {
        employees, employeeFormModal, employee
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchAllEmployees1: () => dispatch(fetchAllEmployeesA()),
        employeeFormMoalOpen1: () => dispatch(employeeFormMoalOpenA()), // TP
        employeeFormMoalClose1: () => dispatch(employeeFormMoalCloseA()),
        addEmployee1: (data) => dispatch(addEmployeeA(data)),
        fetchEmployee1: (empId) => dispatch(fetchEmployeeA(empId)),
        updateEmployee1: (data) => dispatch(updateEmployeeA(data)),
        deleteEmployee1: (empId) => dispatch(deleteEmployeeA(empId)),
        deleteMultipleEmployees1: (empIdArr) => dispatch(deleteMultipleEmployeesA(empIdArr))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EmployeePage);