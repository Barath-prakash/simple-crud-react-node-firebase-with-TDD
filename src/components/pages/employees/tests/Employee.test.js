import React from 'react';
import ReactDOM from 'react-dom';
// Installed plugins 
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';

// Combine reducer file
import rootReducer from 'rootReducer';
import { wrap } from 'module';

import renderer from 'react-test-renderer';
import jest from 'jest';
import { shallow, mount, render } from 'enzyme';
import sinon from 'sinon';

//Components
import EmployeePage from '../EmployeePage';
import EmployeeList from '../EmployeeList';
import FormModal from 'components/forms/employees/FormModal';

// Redux-mock-store
import configureStore from 'redux-mock-store';
// Redux store config
const store = createStore(rootReducer, applyMiddleware(thunk));

it('Console error check', () => {
    const consoleError = sinon.spy(global.console, 'error'); 
    // console.log(consoleError);
    // expect(consoleError).toBe(null);
})

describe("Parent component full test", () => {
    let empComponent;
    beforeEach(() => {
        // const actions = {fetchAll: jest.fn()}
        empComponent = mount(<Provider store={store}><EmployeePage /></Provider>, { lifecycleExperimental: true });
    });

    it("Function calls test", () => {
        // Snapshot check
            const tree = renderer.create(empComponent).toJSON();
            expect(tree).toMatchSnapshot();
        
            // expect(empComponent).toBeDefined();
        // createPortal
            let createPortalFn = ReactDOM.createPortal = (element, node) => element;
    
        // Mounting check
            // expect(empComponent.instance()._isMounted).toBe(true); Nu
        
        // Childern existance check
            // let empMainNode = empComponent.find('.container');
            // expect(empMainNode.children('EmployeeList').length).toEqual(1); Nu
            // expect(empMainNode.children('FormModal').length).toEqual(1); Nu
        
        // DidMount call check
            const didMountLf = sinon.spy(empComponent.instance(), 'componentDidMount');
            const renderFn = sinon.spy(empComponent.instance(), 'render');
            // console.log(didMountLf.calledImmediatelyAfter(renderFn)); 
            // expect(didMountLf).toBeCalled();
            // console.log(didMountLf, renderFn)
        
        let empPage = empComponent.find('EmployeePage');
        // FetchAllEmp function
            const mockFetchEmployeesFn = sinon.spy(empPage.instance(), 'fetchAllEmployees');
            // console.log(mockFetchEmployeesFn())
            // expect(mockFetchEmployeesFn.called).toBe(true); // return false
    
        // Refs
            const setState = sinon.spy(empPage.instance(), 'setState');
            // expect(setState.called).toBe(true); // return false
       
        // Render functions 
            expect(empPage.instance().employeeFormMoalOpen()).toBeDefined(); 
            expect(empPage.instance().employeeFormMoalClose()).toBeDefined();
            expect(empPage.instance().addEmployee()).toBeDefined();
            expect(empPage.instance().editEmployeeP()).toBeDefined();
            expect(empPage.instance().updateEmployee()).toBeDefined();
            expect(empPage.instance().deleteEmployee()).toBeDefined();
            expect(empPage.instance().deleteMultipleEmployeesC()).toBeDefined();
            
            // console.log(empPage.instance().notifyMsgFn())
            // console.log(empComponent.instance())
        
        // Dispatch render props check
            //expect(empPage.renderProp('fetchAllEmployees1')).toBeTruthy(); Nu
    
        // Reducer props type check
            // let empPageProps = empPage.props();
            // expect(empPageProps.employees).toEqual(Array()); Nu
            // expect(empPageProps.employeeFormModal).toEqual(Boolean()); Nu
            // expect(empPageProps.employee).toEqual(Object()); Nu
            
        // State variable type check Nu
            // expect(empPage.instance().state.mDeleteEmpArr).toBeDefined(); 
        
            empComponent.unmount();
    });
})

describe("EmployeeList child component test", () => {

    it("Function calls test", () => {
        const empListComponent = shallow(<EmployeeList />).dive();
        // console.log(empListComponent.find("#empList").props());
        console.log(empListComponent.find(".chkInput"))
       // expect(empListComponent.find(".chkInput").simulate('change', { preventDefault () {} })).toBeDefined();
       // expect(empListComponent.find('Button').simulate('click', { preventDefault () {} })).toBeDefined();
        empListComponent.unmount();
    });
});

describe("FormModal child component test", () => {

    it("Form and functionality test", () => {
        const formComponent = shallow(<FormModal />, { lifecycleExperimental: true });
        
        // Component to be defined Nu
            // expect(formComponent).toBeDefined();
        
        // Form modal existance check based on className Nu
            // expect(formComponent.find('#modal').length).toEqual(1);
        
        // Component state obj check Nu
            // let stateObj = {"empId": "", "firstName": "", "lastName": "", "email": "", "gender": "", "dateOfBirth": "", "age": ""};
            // expect(formComponent.state('data')).toEqual(stateObj);
        
        // Form existance check Nu
            // expect(formComponent.find('form').exists()).toBe(true);
        
        // Form fields existance check
            let firstName = formComponent.find('#firstName'), lastName = formComponent.find('#lastName'), email = formComponent.find('#email'),
                gender = formComponent.find('#gender'), genderMaleFemale = formComponent.find('#gender-male' && '#gender-female'),
                datepicker = formComponent.find('#datepicker'), age = formComponent.find('#age');
            // expect(firstName.length).toBe(1); Nu all
            // expect(lastName.length).toBe(1);
            // expect(email.length).toBe(1);
            // expect(gender.length).toBe(1);
            // expect(datepicker.length).toBe(1);
            // expect(age.length).toBe(1);
                
            firstName.simulate("change", { target: { value: "myValue" } }); expect(firstName.props().value).toBeDefined();
            lastName.simulate("change", { target: { value: "myValue" } }); expect(lastName.props().value).toBeDefined();
            email.simulate("change", { target: { value: "myValue" } }); expect(email.props().value).toBeDefined();
            genderMaleFemale.simulate("change", { target: { value: "myValue" } }); expect(genderMaleFemale.props().value).toBeDefined();
            datepicker.simulate("change", { target: { value: "myValue" } }); expect(datepicker.props().selected).toBeDefined();
            age.simulate("change", { target: { value: "myValue" } }); expect(age.props().value).toBeDefined();

        // Submit event trigger check
            // expect(formComponent.find('form').simulate('click', { preventDefault () {} })).toBeDefined(); // Nu
            expect(formComponent.find('form').simulate('submit', { preventDefault () {} })).toBeDefined();
            

        // Handle function check Nu
            // let formInstance = formComponent.instance();
            // expect(formInstance.handleCloseFormModal).toBeInstanceOf(Function);
            // expect(formInstance.handleChange).toBeInstanceOf(Function);
            // expect(formInstance.handleDateChange).toBeInstanceOf(Function);
            // expect(formInstance.handleSubmit).toBeInstanceOf(Function);
            // expect(formInstance.validate).toBeInstanceOf(Function);
            // expect(formInstance.resetForm).toBeInstanceOf(Function);
        
        // Didupdate call check  Nu
            // const didUpdateLf = sinon.spy(formComponent.instance(), 'componentDidUpdate');
            // expect(didUpdateLf.calledAfter.bind()).toBeDefined();
        
            formComponent.unmount();
    });
});
