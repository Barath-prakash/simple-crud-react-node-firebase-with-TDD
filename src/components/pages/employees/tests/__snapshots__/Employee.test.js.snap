// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parent component full test Function calls test 1`] = `
<div
  className="container"
>
  <div
    className="notifications-wrapper"
    style={Object {}}
  />
  <div>
    <h4
      style={
        Object {
          "float": "left",
        }
      }
    >
      Employees List
    </h4>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "float": "right",
        }
      }
      type="button"
    >
      Add employee
    </button>
  </div>
  <br />
  <br />
  <table
    className="table table-striped table-bordered table-hover"
    id="empList"
  >
    <thead>
      <tr>
        <th>
          Multiple delete
        </th>
        <th>
          Employee name
        </th>
        <th>
          Email
        </th>
        <th>
          Gender
        </th>
        <th>
          DOB
        </th>
        <th>
          Age
        </th>
        <th>
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colSpan="7"
        >
          No employees found
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
